activerecord/lib/active_record/enum.rb

```diff
@@ -138,16 +138,12 @@ def deserialize(value)
         mapping.key(subtype.deserialize(value))
       end
 
-      def serializable?(value)
-        (value.blank? || mapping.has_key?(value) || mapping.has_value?(value)) && super
-      end
-
       def serialize(value)
         mapping.fetch(value, value)
       end
 
       def assert_valid_value(value)
-        unless serializable?(value)
+        unless value.blank? || mapping.has_key?(value) || mapping.has_value?(value)
```

Comment:
Can you instead remove the `super` from `serializable?`?

Reply:
@eileencodes  I've tried to [restore serializable?](https://github.com/yevgenko/rails/commit/b7fafb28c493118d7dbe6b401e0d97e1dd751224), but test has failed.
Looks like removing `super` from `serializable?` alone can't fix the problem, but removing `serializable?` can.
<details>
<summary>test failed</summary>

```
Using sqlite3
Run options: --seed 20368

# Running:

.........F

Failure:
EnumTest#test_find_via_where_with_values.to_s [/vagrant/rails/activerecord/test/cases/enum_test.rb:69]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Book id: 1, author_id: 1, format: "paperback", name: "Agile Web Development with Rails", status: "published", last_read: "read", nullable_status: nil, language: "english", author_visibility: "visible", illustrator_visibility: "visible", font_size: "medium", difficulty: "medium", cover: "hard", isbn: nil, external_id: nil, published_on: nil, boolean_status: "enabled", tags_count: 0, created_at: "2021-02-19 16:30:24.429872000 +0000", updated_at: "2021-02-19 16:30:24.429872000 +0000", updated_on: "2021-02-19">
+nil



bin/test test/cases/enum_test.rb:64

...................................................................

Finished in 2.017204s, 38.1716 runs/s, 126.9083 assertions/s.
77 runs, 256 assertions, 1 failures, 0 errors, 0 skips
```

</details>

Reply:
Alternatively we could cast stringified digits into numeric type, for example:

0ffa6e51d286e2d6e850c0a054b1d26cfdd2975e

What do you think? :)

Reply:
@yevgenko It will break the existing behavior effed76ec6d7863ff59fb2df02508ce6202498a9.

> Can you instead remove the `super` from `serializable?`?

I'm not sure I get your point, so let me explain why the `value.blank? || mapping.has_key?(value) || mapping.has_value?(value)` part should not be left in the `serializable?`.

`EnumType#serialize` is `mapping.fetch(value, value)`, it means that all values were serializable before 72fd0bae5948c1169411941aeea6fef4c58f34a9.

https://github.com/rails/rails/blob/effed76ec6d7863ff59fb2df02508ce6202498a9/activerecord/lib/active_record/enum.rb#L146-L148

After 72fd0bae5948c1169411941aeea6fef4c58f34a9, `serializable?` is introduced which checks a value is whether `cast`-able or not, it is used only in `HomogeneousIn`.

https://github.com/rails/rails/blob/effed76ec6d7863ff59fb2df02508ce6202498a9/activerecord/lib/arel/nodes/homogeneous_in.rb#L50-L52

So now, all values are no longer always serializable in `HomogeneousIn`, it is the cause for the issue #41474.

I'm not sure about we'd like the enum serialization more stricter in the future, at least for now the serialization inconsistency whether in `HomogeneousIn` or others should be fixed.